In [1]:
import pandas as pd
import math
import numpy as np

In [3]:
def factorial(x):
    if x == 1:
        return 1
    else:
        return x*factorial(x-1)

In [5]:
factorial(3)

6

In [6]:
factorial(0)

RecursionError: maximum recursion depth exceeded in comparison

In [7]:
math.factorial(0)

1

In [11]:
def true_factorial(x):
    
    def factorial(x):
        if x == 1:
            return 1
        else:
            return x*factorial(x-1)
    
    return int(factorial(x+1)/(x+1))
    

In [15]:
true_factorial(5)

120

n! is the number of ways we can arrange n objects

$$ x! = x \times (x-1) \times (x-2) \times ... \times 2 \times 1 $$ 

$$ x! = \frac{(x+1)!}{(x+1)} = \frac{(x+1) \times x \times (x-1) \times (x-2) \times ... \times 2 \times 1}{x+1} $$ \
$$ 4! = \frac{5!}{5} = \frac{120}{5} = 24 $$ \
$$ 3! = \frac{4!}{4} = \frac{24}{4} = 6 $$ \
$$ 2! = \frac{3!}{3}  = \frac{6}{3} = 2  $$ \
$$ 1! = \frac{2!}{2} = \frac{2}{2} = 1 $$ \
$$ 0! = \frac{1!}{1} = \frac{1}{1} = 1 $$